src/Utils/Logging/Trace.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Module: Utils.Logging.Trace
+-- Copyright: Copyright © 2019 Kadena LLC.
+-- License: MIT
+-- Maintainer: Lars Kuhtz <lars@kadena.io>
+-- Stability: experimental
+--
+-- TODO
+--
+module Utils.Logging.Trace
+( trace
+, Trace
+) where
+
+import Control.DeepSeq
+import Control.Monad.IO.Class
+import Control.StopWatch
+
+import Data.Aeson
+import qualified Data.Text as T
+
+import GHC.Generics
+
+import System.Clock
+import System.LogLevel
+
+-- internal modules
+
+import Chainweb.Time
+
+import Data.LogMessage
+
+-- -------------------------------------------------------------------------- --
+-- Logging
+
+data Trace = Trace
+    { _traceAction :: !T.Text
+    , _traceParam :: !Value
+    , _traceWeight :: !Int
+    , _traceTime :: !Micros
+    }
+    deriving (Show, Eq, Generic, NFData)
+
+instance ToJSON Trace where
+    toJSON o = object
+        [ "action" .= _traceAction o
+        , "param" .= _traceParam o
+        , "weight" .= _traceWeight o
+        , "time" .= _traceTime o
+        ]
+
+trace
+    :: MonadIO m
+    => ToJSON param
+    => LogFunction
+    -> T.Text
+    -> param
+    -> Int
+    -> m a
+    -> m a
+trace logg label param weight a = do
+    (!r, t) <- stopWatch a
```

Comment:
This could be expensive to add to every call. I have a couple of proposed knobs to add: first, make it #ifdef-ably no-op with a cabal flag. Second, allow for probabilistic tracing with hash-sampling:

  - add a `mwc-random` RNG to the trace function
  - add an argument to trace `sample :: Word32` -- if sample is zero, sample always, but when nonzero only collect a trace sample when random `x <= sample`.

Reply:
I would have used CPP+cabal flag, but I think, there is a policy in place, that usage of CPP and, generally, conditional compilation via cabal flags should be avoided.

Regarding the probabilistic sampling, I think, that would be the job of a more general profiling system. To me even calls to the PRNG + conditional branch in the code seem too much overhead to be used broadly in the code.

I'd rather keep this stuff very simple and discourage usage for anything but expensive, rage, high-level calls (the calls in the examples all already trigger several other log messages and involve expensive IO). Or make clear that it should be used only on temporary code branches for debugging.

Reply:
I guess, I should add a big warning to the is function, that clarifies how it should be used.

Reply:
I wonder if it would make sense to add your proposed optional probabilist sampling to the `LogFunction` type in general.

That could be an interesting feature for a logger, that would be orthogonal to log-levels for controlling logging volume.